### PR TITLE
K nearest search in quadtree

### PR DESCRIPTION
**Pull request by gardster**
K nearest visitor implementation for the quad tree.

### CODE REVIEW

File: quadtree/quadtree.go [hunk 1]
```diff
@@ -251,6 +259,36 @@ func (q *Quadtree) FindMatching(p *geo.Point, f Filter) geo.Pointer {
...
+	}
+
+	newVisit(v).Visit(q.root,
+		q.bound.Left(), q.bound.Right(),
+		q.bound.Bottom(), q.bound.Top(),
+	)
+
+	//repack result
+	result := make([]geo.Pointer, 0, k)
+	for _, element := range v.closest {
```
Comment (mkabischev) [inline quadtree/quadtree.go]: I suggest to change it to:
```go
result := make([]geo.Pointer, k)
for i, element := range v.closest {
    result[i] = element.point
}
```

File: quadtree/quadtree.go [hunk 2]
```diff
@@ -382,6 +420,93 @@ func (v *findVisitor) Visit(p geo.Pointer) {
 	return
 }
 
+type pointsQueueItem struct {
+	point    geo.Pointer
+	distance float64 // distance to point and priority inside the queue
+	index    int     // point index in queue
+}
+
+type pointsQueue []*pointsQueueItem
```
Comment (mkabischev) [inline quadtree/quadtree.go]: Why do you use pointers here?

File: quadtree/quadtree.go [hunk 3]
```diff
@@ -382,6 +420,93 @@ func (v *findVisitor) Visit(p geo.Pointer) {
...
+
+			// We have filled queue, so we start to restrict searching range
+			d = math.Sqrt(top.distance)
+			x := v.point.X()
+			y := v.point.Y()
+			v.closestBound.Set(x-d, x+d, y-d, y+d)
+		}
+	}
+
+	return
```
Comment (sbinq) [inline quadtree/quadtree.go]: Is this `return` necessary?

Comment (gardster) [reply quadtree/quadtree.go]: similar with previous `Visitor` implementation. Fixed both.

File: quadtree/quadtree.go (hunk 2 above)
Comment (gardster) [reply quadtree/quadtree.go]: Rewritten to values

File: quadtree/quadtree.go [hunk 4]
```diff
@@ -378,8 +416,91 @@ func (v *findVisitor) Visit(p geo.Pointer) {
...
+
+type pointsQueueItem struct {
+	point    geo.Pointer
+	distance float64 // distance to point and priority inside the queue
+	index    int     // point index in queue
+}
+
+type pointsQueue []pointsQueueItem
+
+func NewPointsQueue(capacity int) pointsQueue {
```
Comment (paulmach) [inline quadtree/quadtree.go]: This method should not be exported. Push expects a `pointsQueueItem` which isn't exported.

Comment (gardster) [reply quadtree/quadtree.go]: Yes, it shouldn't. Fixed.

**Comment by paulmach** [comment]
Thank you for the contribution!!

**Comment by gardster** [comment]
Thank you for the fast review!

**Pull request merged**